Null check in ResourceRequest constructor

nmahendru opened the pull request:
#### 1372d75734871f8100fdd697921487bffafcbc07
<pre>
Null check in ResourceRequest constructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=282201">https://bugs.webkit.org/show_bug.cgi?id=282201</a>
<a href="https://rdar.apple.com/137742249">rdar://137742249</a>

Reviewed by Alex Christensen and David Kilzer.

This is just a sanity check for the optional field m_requestor.

* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::ResourceRequest):

Canonical link: <a href="https://commits.webkit.org/285920@main">https://commits.webkit.org/285920@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a09782442a3f1fdf79100025fea6e470eee48636

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73867 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53296 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26678 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78198 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25105 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75984 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1081 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58081 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16435 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48229 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63556 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38476 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45037 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21045 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23438 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66594 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21393 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79756 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1184 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/609 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66408 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1327 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63569 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65688 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16341 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9582 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7764 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1148 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3898 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1177 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1164 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1183 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
```diff
@@ -64,7 +64,8 @@
+        if (platformData.m_requester)
+            setRequester(*platformData.m_requester);
         m_nsRequest = platformData.m_urlRequest;
         setIsAppInitiated(*platformData.m_isAppInitiated);
```
The `m_isAppInitiated` field is also `std::optional<>`, so this could also crash due to `platformData.m_isAppInitiated` not being set:

```
    std::optional<bool> m_isAppInitiated;
```

One way to fix this one is to use `std::optional<>::value_or()` (assuming `false` should be the default value if the `std::optional<>` is not set):

```suggestion
        setIsAppInitiated(platformData.m_isAppInitiated.value_or(false));
```

You don't have to fix this with this PR, but please fix this in a follow-up as we can't assume a `std::optional<>` is always set like this.

ddkilzer commented on Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
```diff
@@ -64,7 +64,8 @@
 {
     if (platformData.m_urlRequest) {
-        setRequester(*platformData.m_requester);
+        if (platformData.m_requester)
+            setRequester(*platformData.m_requester);
```
Should we provide a default value if `platformData.m_requester` is not set, like this?

```
        setRequester(*platformData.m_requester.value_or(ResourceRequestRequester::Unspecified);
```

Please talk to an expert to see if we should provide a default value here.

nmahendru replied:
That's a good point. But I think that's already happening as the default value would be the first value in the enum(ResourceRequestRequester::Unspecified)

achristensen07 replied:
WebCore::ResourceRequestRequester::Unspecified is already the default value, so that seems unnecessary.

ddkilzer reviewed (approved):
r=me, but please ask @achristensen07 or another expert if we should use `ResourceRequestRequester::Unspecified` if no value is set in the `std::optional<>`.

achristensen07 reviewed (approved):
We may want to consider using a std::optional<std::tuple<RetainPtr<NSURLRequest>, bool, ResourceRequestRequester>> instead.

nmahendru committed:
Null check in ResourceRequest constructor
https://bugs.webkit.org/show_bug.cgi?id=282201
rdar://137742249

Reviewed by Alex Christensen and David Kilzer.

This is just a sanity check for the optional field m_requestor.

* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::ResourceRequest):

Canonical link: https://commits.webkit.org/285920@main
